Add filters `|default`, `|assigned_or` and `|defined_or`

Kijewski opened the pull request:
Also, `enum Pluralize<S, P>` is renamed into `enum Either<L, R>` and exported.

Cc @joshka.

Resolves <https://github.com/askama-rs/askama/issues/424>.

GuillaumeGomez commented on askama_derive/src/generator/filter.rs:
```diff
@@ -643,6 +706,8 @@ fn ensure_filter_has_feature_alloc(
     Ok(())
 }
 
+const FALSE: &WithSpan<'static, Expr<'static>> = &WithSpan::new_without_span(Expr::BoolLit(false));
```
Considering it's only used by `visit_default_filter`, maybe it'd be better to move it there?

Kijewski replied:
It's also used in `visit_indent_filter`.

GuillaumeGomez commented:
Btw, no clue about a better for `DefaultUnwrappable` so I'm fine with it.

joshka commented on book/src/filters.md:
````diff
@@ -78,6 +78,59 @@ Output:
+</li><li>
+
+```jinja
+{{ variable_or_expression | default(default_value, true) }}
````
This sort of parameter without a name makes it difficult to understand a template worth reading the template docs. I'd been inclined to either:
1. Replace a naked bool parameter with an enum that self documents (that may look verbose here), or
2. Replace a bool with a second named method. E.g. default_if_undefined might be a good name for the false branch. Alternatively default_if_none / err

Kijewski replied:
I added a line that `{{ variable_or_expression | default(default_value, boolean = true) }}` works, too.

joshka replied:
I'd still prefer explicit over implicit a bit. This is based on an assumption that the direction of askama is a jinja-like template library with natural rust idioms rather than a jinja compatible template library that adds rust conveniences. I'm not sure whether that assumption is correct however (intuitively I'd lean towards the former as being more useful)

As a single data point. Having not seen the jinja syntax for the `default` filter prior, even with the parameter named `boolean` in the CoPilot suggestion rather than the bare `true`, I still had no idea from reading it what that parameter meant or controlled seeing it for the first time. This suggests to me that it is poorly named in jinja, and something that we could make better pretty easily with good naming here.

joshka commented:
> Also, it would be possible to implement the trait for integers (testing if != 0), and Strings (testing if != "").

Empty string seems reasonable as that's rationally returned when a string is not entered somewhere, but most of the time a zero value seems like it would be only 0 if you're using that as a flag value, which is not very rusty, so I'd perhaps implement default_if_empty for strings (and str, cow perhaps), but not default_if_zero unless there's an explicit obvious use case where it makes sense.

Kijewski commented:
* Renamed trait to `DefaultFilterable` (I think that's better)
* Implemented for `Cow<'_, T>` (I forgot about taat, good call!)
* Implemented for `Wrapping<T>` & `Saturating<T>`
* Implemented for `NonZero*` (infallably `Ok(Some())`) [just so you don't have to update the code if you use NonZeroU32 and u32]

I think implementing the type for integers is fine and useful. In many real-world cases you will find `0` being used as a poor-man's `NULL`. And unless `boolean = true`, it does not matter either way. `DefaultFilterable` is not recursive: `Some(0)` returns `0`; only the `Option` is evaluated, not the contained value.

GuillaumeGomez commented on askama/src/filters/builtin.rs:
````diff
@@ -330,13 +331,13 @@ impl<T: fmt::Display> fmt::Display for Center<T> {
 /// # }
 /// ```
 #[inline]
-pub fn pluralize<C, S, P>(count: C, singular: S, plural: P) -> Result<Pluralize<S, P>, C::Error>
+pub fn pluralize<C, S, P>(count: C, singular: S, plural: P) -> Result<Either<S, P>, C::Error>
````
Well, version bumping will be needed since we plan to extend `{% call %}` blocks. So it's fine.

Kijewski replied:
The type was unreachable before, so its name did not matter.

GuillaumeGomez commented on askama/src/filters/builtin.rs:
```diff
@@ -426,31 +427,270 @@ const _: () = {
+    fn as_filtered(&self) -> Result<Option<Self::Filtered<'_>>, Self::Error>;
+}
+
+const _: () = {
```
To prevent impacting the upper scope.

Kijewski replied:
In here I mostly use it like a [`#region`](https://stackoverflow.com/questions/44390454/what-is-the-use-of-region-and-endregion-in-c). Every `DefaultFilterable` is in this scope and one click hides everything.

GuillaumeGomez commented on askama/src/filters/builtin.rs:
```diff
@@ -426,31 +427,270 @@ const _: () = {
+
+        #[inline]
+        fn as_filtered(&self) -> Result<Option<&T>, Infallible> {
+            match self {
+                Ok(value) => Ok(Some(value)),
+                Err(_) => Ok(None),
+            }
```
Nice idea!

GuillaumeGomez commented on askama_derive/src/generator/filter.rs:
```diff
@@ -627,6 +626,70 @@ impl<'a> Generator<'a, '_> {
+        let Expr::BoolLit(do_unwrap) = **boolean else {
+            return Err(ctx.generate_error(
+                format_args!(
+                    "the `default` filter takes a boolean literal as it's optional second argument"
```
```suggestion
                    "the `default` filter takes a boolean literal as its optional second argument"
```

GuillaumeGomez commented on askama_derive/src/generator/filter.rs:
```diff
@@ -627,6 +626,70 @@ impl<'a> Generator<'a, '_> {
+            let Some(var_name) = var_name else {
+                return Err(ctx.generate_error(
+                    format_args!(
+                        "the `default` filter requires a variable name as on its left-hand side"
```
"as on"?

GuillaumeGomez commented on askama_derive/src/generator/filter.rs:
```diff
@@ -627,6 +626,70 @@ impl<'a> Generator<'a, '_> {
+                default_value: None,
+            },
+            &FilterArgument {
+                name: "boolean",
```
Why not naming it "do_unwrap" or something along the line to better match what it's doing instead?

GuillaumeGomez replied:
Question still open. :)

Kijewski replied:
The filter only makes sense if it behaves the same as in jinja.

GuillaumeGomez replied:
That saddens me, but it's a valid point.

Kijewski commented:
So, what do we want to do with the argument "boolean"? It's the name and semantics Jinja uses. The question is: I compatible do we want to be?

If we don't care to be 100% compatible (and we don't), then yes, two filters would be the better option:
1. I would then call the `boolean = false` option `{{ var | defined_or(default_value) }}`.
2. And the `boolean = true` option `{{ expr | assigned_or(default_value) }}`.

Fun fact minijinja does not implement `|default` either.

Kijewski commented:
Or, what about _both_ options?
* `{{ var | default(default_value, [false]) }}` → `{{ var | defined_or(default_value) }}`
* `{{ expr | default(default_value, [boolean =] true) }}` → `{{ expr | assigned_or(default_value) }}`

The filter `|default` will work as in jinja, same semantics, name argument naming; but you could also use filters with a (hopefully) better naming scheme.

GuillaumeGomez commented:
Having two filters for the same thing sounds good to me. :+1:

Just please improve the argument name. We can add an error for this specific case to tell that for better naming, `askama` is using `X` name instead for this named argument.

What do you think @joshka?

joshka commented:
Supporting ninja syntax probably makes it easier to bring templates from elsewhere (and allows LLMs to do fall into the pit of success). So while I dislike the syntax, if there's a better syntax available like suggested above, then I've no complaints about providing compatible syntax like this. I'd even go so far as to suggest that the parameter name is fine in that case (alternatively, if there's some future ability to alias it to give it an additional more descriptive name then that would work too)

Kijewski commented:
I added `|assigned_or` and `|defined_or` as (better) alternatives to `|defined`.

I don't think that it would be a good idea to implement `|defined` with other arguments than in Jinja.
* EITHER we keep it for Jinja compatibility exactly like in jinja
* OR we drop it, because the other two filters are better
* OR we re-implement the `|default` filter to `compile_error!` with a message to use the other filters.

joshka commented:
The other names are good.

I like the idea of making it possible to do something useful if someone tries jinja syntax. That makes a lot of sense as it makes it easy to fall into a pit of success.

But I also like the idea of making this a compile error, but wonder if it might be possible to cause a deprecation warning to be emitted for `default` instead? Are there cases where a template might want to need default|undefined type logic in askama? I.e. is there some case where a template would call another template but intentionally not provide some field?

Kijewski commented:
Blocked by <https://github.com/askama-rs/askama/pull/471>.

----

> Are there cases where a template might want to need default|undefined type logic in askama? I.e. is there some case where a template would call another template but intentionally not provide some field?

I could not think of a use case. In python, the `boolean` argument can be evaluated at runtime, but with our pre-compiled code this is not possible.

Right now, a proc-macro can only emit fatal errors, no warnings. There are open RFCs to add `compile_warning!`-like macros, though. But I guess we could always add a call to ↓ to fake a deprecation message.
```rust
#[deprecated(note = "use `|defined_or` or `|assigned_or` instead of `|default`")]
fn dont_use_default_default() {}
```
Most warnings that come from proc-macros get suppressed, though. I would need to test if deprecation message are one of them.

Kijewski commented:
We can't assign a span to the warning (→ <https://github.com/askama-rs/askama/issues/420>). Without the span, such a warning is kinda useless IMHO. I think the current implementation is fine as it is. :) We can always add a warning in a subsequent PR. @GuillaumeGomez, @joshka, what do you think? Good enough to be merged?

GuillaumeGomez commented:
Output looks ok to me.

joshka commented on askama/src/filters/builtin.rs:
```diff
@@ -426,31 +427,265 @@ const _: () = {
+    label = "`{Self}` is not `|assigned_or` filterable",
+    message = "`{Self}` is not `|assigned_or` filterable"
+)]
+pub trait DefaultFilterable {
```
It could be worth adding a unit tests for the various implementations of this that captures each implementation. E.g.:

```rust
assert_eq!(0_u8.as_filtered(), Ok(None));
assert_eq!(1_u8.as_filtered(), Ok(Some(1));
```

joshka commented on book/src/filters.md:
````diff
@@ -27,6 +27,30 @@ which are `singular = ""` and `plural = "s"` by default.
+
+```jinja
+{% let greeting = Some("Hello") %}
+{{ greeting.as_ref() | default("Hi", true) }}
````
should be 
```suggestion
{{ greeting.as_ref() | assigned_or("Hi") }}
```

joshka commented on book/src/filters.md:
````diff
@@ -27,6 +27,30 @@ which are `singular = ""` and `plural = "s"` by default.
+[#assigned_or]: #assigned_or
+
+```jinja
+{{ variable_or_expression | assigned_or(fallback) }}
+{{ variable_or_expression | assigned_or(fallback) }}
````
duplicate line - needs to be fixed or state what's going on here.

joshka commented on book/src/filters.md:
````diff
@@ -78,6 +102,46 @@ Output:
+{{ variable_or_expression | default(default_value, [[boolean =] true]) }}
+```
+
+This filter works like the Jinja filter of the same name.
+If the second argument is not an boolean `true`, then the filter behaves like [`|defined_or`][#defined_or].
+If it is supplied and `true`, then the filter behaves like [`|assigned_or`][#assigned_or].
````
This should maybe describe how this works without having to link to the jinja / the other filters.

joshka commented on book/src/filters.md:
```diff
@@ -78,6 +102,46 @@ Output:
+If it is supplied and `true`, then the filter behaves like [`|assigned_or`][#assigned_or].
+
+**This filter exists for compatibility with Jinja.
+You should use [`|defined_or`][#defined_or] and [`|assigned_or`][#assigned_or] directly instead.**
```
```suggestion
Askama provides [`|defined_or`][#defined_or] and [`|assigned_or`][#assigned_or] which both better express the intention and should generally usually be used instead of this filter.**
```

joshka commented on testing/tests/ui/default.stderr:
```diff
@@ -0,0 +1,88 @@
+       "var | default }}"
+  --> tests/ui/default.rs:10:35
+   |
+10 | #[template(ext = "html", source = "{{ var | default }}")]
```
This actually suggests a bit of a jinja / rust impedance mismatch here. In Rust, we have a few different operations like `map_or_default` and `unwrap_or_default`. Intuitively, `var | default` when the type of var implements the `Default` trait seems like it would be obvious that it should return default. I know that a user could call `.unwrap_or_default()` or whatever there, but I'm curious whether it might be worth extending this to allow the obvious intuitive thing there. (I'm not fully certain of this, either way, WDYT?)

joshka reviewed (commented):
Some extra rview notes having gone through this in a bit more detail.

Kijewski commented:
@joshka, thank you for your suggestions and proofreading! I will have a look tomorrow.

Kijewski commented:
Took me a little longer than a day to look into it. 😳

GuillaumeGomez commented on askama/src/filters/default.rs:
```diff
@@ -0,0 +1,403 @@
+    )* };
+}
+
+impl_for_non_zero!(
```
Since we need to provide both the non-zero ty and the integer ty, why not merge this macro with `impl_for_int`? Like that: less code and one less macro definition/expansion.

GuillaumeGomez commented on askama/src/filters/default.rs:
```diff
@@ -0,0 +1,403 @@
+/// not sub-normal, not infinite and not NaN.
+impl DefaultFilterable for f32 {
+    type Filtered<'a>
+        = Self
```
Wow this formatting is super weird. O.o

Kijewski replied:
Yeah, it's not the formatting I would have preferred, but rustfmt insists on adding a linebreak there. :)

GuillaumeGomez commented on askama/src/filters/default.rs:
```diff
@@ -0,0 +1,403 @@
+
+/// An `f64` has a value if it is [`Normal`](FpCategory::Normal), i.e. it is not zero,
+/// not sub-normal, not infinite and not NaN.
+impl DefaultFilterable for f64 {
```
Considering it's the same implementation as `f32` and that rust already has `f16` and `f128` in nightly, should we put this in a macro so we can more easily implement it on other floating types?

Kijewski replied:
Good idea! Made it a macro call.

GuillaumeGomez commented on testing/tests/default.rs:
```diff
@@ -0,0 +1,594 @@
+
+#[test]
+fn test_default_with_str() {
+    // In this test a default value should be returned for an empty string.
```
Comments explaining what the test checks is very appreciated, thanks! :)

GuillaumeGomez commented on testing/tests/ui/default.stderr:
```diff
@@ -0,0 +1,88 @@
+40 | #[template(ext = "html", source = "{{ var | defined_or }}")]
+   |                                   ^^^^^^^^^^^^^^^^^^^^^^^^
+
+error[E0277]: `Mutex<u32>` is not `|assigned_or` filterable
```
I didn't know we could do that in rust, it looks super nice! (and will look even better once I add the spans to the generated `TokenStream` :3)

Kijewski replied:
The feature is already stable for over a year: <https://releases.rs/docs/1.78.0/>, but I am not sure if I ever saw it in the wild. I found it [in the standard library](https://doc.rust-lang.org/1.88.0/src/core/future/future.rs.html#32) and simply tested if library developers can use it, too. :)

GuillaumeGomez replied:
A lot of features are, but if I don't use them, I simply forget about them. ^^'

GuillaumeGomez commented:
Looks all good to me, thanks!
